File: 2.C++_Calculator/core/CalculatorEngine.cpp
#include "CalculatorEngine.h"
#include <iostream>

using namespace std;

CalculatorEngine::CalculatorEngine() {
    lastResult = 0.0;
}

// Basic arithmetic operations
double CalculatorEngine::add(double a, double b) {
    return a + b;
}

double CalculatorEngine::subtract(double a, double b) {
    return a - b;
}

double CalculatorEngine::multiply(double a, double b) {
    return a * b;
}

double CalculatorEngine::divide(double a, double b) {
    if (b == 0) {
        cout << "Error: Division by zero!" << endl;
        return 0;
    }
    return a / b;
}

double CalculatorEngine::performBasicOperation(double a, double b, char operation) {
    double result = 0.0;
    
    switch (operation) {
        case '+':
            result = add(a, b);
            break;
        case '-':
            result = subtract(a, b);
            break;
        case '*':
            result = multiply(a, b);
            break;
        case '/':
            result = divide(a, b);
            break;
        default:
            cout << "Invalid operation!" << endl;
            return 0;
    }
    
    lastResult = result;
    return result;
}

// Trigonometric operations
double CalculatorEngine::performTrigOperation(const string& function, double value) {
    double result = 0.0;
    
    if (function == "sin" || function == "sine") {
        result = trigCalc.sine(value);
    } else if (function == "cos" || function == "cosine") {
        result = trigCalc.cosine(value);
    } else if (function == "tan" || function == "tangent") {
        result = trigCalc.tangent(value);
    } else if (function == "asin" || function == "arcsine") {
        result = trigCalc.arcsine(value);
    } else if (function == "acos" || function == "arccosine") {
        result = trigCalc.arccosine(value);
    } else if (function == "atan" || function == "arctangent") {
        result = trigCalc.arctangent(value);
    } else {
        cout << "Invalid trigonometric function!" << endl;
        return 0;
    }
    
    lastResult = result;
    return result;
}

void CalculatorEngine::setAngleMode(bool degrees) {
    trigCalc.setAngleMode(degrees);
}

bool CalculatorEngine::getAngleMode() {
    return trigCalc.getAngleMode();
}

string CalculatorEngine::getAngleModeString() {
    return trigCalc.getAngleMode() ? "Degrees" : "Radians";
}

// Memory operations
void CalculatorEngine::storeInMemory(double value) {
    memory.store(value);
}

double CalculatorEngine::recallFromMemory() {
    return memory.recall();
}

void CalculatorEngine::clearMemory() {
    memory.clear();
}

bool CalculatorEngine::hasMemoryValue() {
    return !memory.isEmpty();
}

void CalculatorEngine::showMemoryStatus() {
    memory.displayStatus();
}

// Result management
double CalculatorEngine::getLastResult() {
    return lastResult;
}

void CalculatorEngine::setLastResult(double result) {
    lastResult = result;
}

// Validation
bool CalculatorEngine::isValidOperation(char operation) {
    return (operation == '+' || operation == '-' || operation == '*' || operation == '/');
}

bool CalculatorEngine::isValidTrigFunction(const string& function) {
    return (function == "sin" || function == "cos" || function == "tan" ||
            function == "asin" || function == "acos" || function == "atan" ||
            function == "sine" || function == "cosine" || function == "tangent" ||
            function == "arcsine" || function == "arccosine" || function == "arctangent");
}

File: 2.C++_Calculator/mobile/lib/calculator_engine.dart
import 'dart:math' as math;

/// Calculator Engine - Dart implementation of C++ calculator core
/// This mirrors the functionality of our C++ CalculatorEngine class
class CalculatorEngine {
  double _memoryValue = 0.0;
  bool _hasMemoryValue = false;
  double _lastResult = 0.0;

  // Basic operations - identical to C++ core
  double add(double a, double b) => a + b;
  double subtract(double a, double b) => a - b;
  double multiply(double a, double b) => a * b;
  
  double divide(double a, double b) {
    if (b == 0) throw ArgumentError('Division by zero');
    return a / b;
  }

  // Trigonometric operations with degree/radian conversion
  double sine(double angle, bool isDegreesMode) {
    final radians = isDegreesMode ? _degreesToRadians(angle) : angle;
    return math.sin(radians);
  }

  double cosine(double angle, bool isDegreesMode) {
    final radians = isDegreesMode ? _degreesToRadians(angle) : angle;
    return math.cos(radians);
  }

  double tangent(double angle, bool isDegreesMode) {
    final radians = isDegreesMode ? _degreesToRadians(angle) : angle;
    return math.tan(radians);
  }

  double arcsine(double value, bool isDegreesMode) {
    if (value < -1 || value > 1) {
      throw ArgumentError('Domain error: Input must be between -1 and 1');
    }
    final result = math.asin(value);
    return isDegreesMode ? _radiansToDegrees(result) : result;
  }

  double arccosine(double value, bool isDegreesMode) {
    if (value < -1 || value > 1) {
      throw ArgumentError('Domain error: Input must be between -1 and 1');
    }
    final result = math.acos(value);
    return isDegreesMode ? _radiansToDegrees(result) : result;
  }

  double arctangent(double value, bool isDegreesMode) {
    final result = math.atan(value);
    return isDegreesMode ? _radiansToDegrees(result) : result;
  }

  // Memory operations - identical to C++ implementation
  void storeInMemory(double value) {
    _memoryValue = value;
    _hasMemoryValue = true;
  }

  double recallFromMemory() {
    return _memoryValue;
  }

  void clearMemory() {
    _memoryValue = 0.0;
    _hasMemoryValue = false;
  }

  bool hasMemoryValue() {
    return _hasMemoryValue;
  }

  // Utility methods
  double getLastResult() => _lastResult;

  void setLastResult(double result) {
    _lastResult = result;
  }

  // Private helper methods for angle conversion
  double _degreesToRadians(double degrees) => degrees * math.pi / 180;
  double _radiansToDegrees(double radians) => radians * 180 / math.pi;
}